File: frontend/src/types/writing-workspace.ts
// ─────────────────────────────────────────────────────────────
// Writing Workspace types
// Mirrors GET/POST /api/v1/writing-workspace/*
// ─────────────────────────────────────────────────────────────

export type WritingTaskType = 'task_1' | 'task_2';

export type WritingSubmissionStatus = 'draft' | 'submitted';

export interface WritingWorkspacePrompt {
  id: string;
  task_type: WritingTaskType;
  title: string;
  prompt_text: string;
  word_limit: number;
  time_limit_seconds: number;
  difficulty: number;
  topics?: string[] | null;
}

export interface WritingWorkspacePromptsResponse {
  task_type: string;
  prompts: WritingWorkspacePrompt[];
  total: number;
}

export interface WritingWorkspacePromptResponse {
  id: string;
  task_type: WritingTaskType;
  title: string;
  prompt_text: string;
  word_limit: number;
  time_limit_seconds: number;
  difficulty: number;
  topics: string[];
}

export interface WritingWorkspaceSubmission {
  id: string;
  user_id: string;
  prompt_id: string | null;
  task_type: WritingTaskType;
  title: string;
  prompt_text: string | null;
  word_limit: number;
  time_limit_seconds: number;
  essay_text: string;
  word_count: number;
  time_seconds_spent: number;
  status: WritingSubmissionStatus;
  is_locked: boolean;
  evaluation_status?: string | null;
  submission_summary: WritingSubmissionSummary;
  created_at: string | null;
  updated_at: string | null;
  submitted_at: string | null;
}

export interface WritingSubmissionSummary {
  word_count: number;
  word_limit: number;
  time_seconds_spent: number;
  time_limit_seconds: number;
  meets_word_requirement: boolean;
  within_time_limit: boolean;
  warnings: string[];
  submitted_at: string | null;
}

export interface WritingWorkspaceSubmissionStart {
  prompt_id: string;
}

export interface WritingWorkspaceSubmissionSave {
  essay_text: string;
  time_seconds_spent?: number;
}

export interface WritingWorkspaceSubmissionSubmit {
  time_seconds_spent?: number;
}

export interface WritingWorkspaceSubmissionListResponse {
  results: WritingWorkspaceSubmission[];
  total: number;
}

export const WRITING_WORKSPACE_TASK_LABELS: Record<WritingTaskType, string> = {
  task_1: 'Academic Task 1',
  task_2: 'Task 2 Essay',
};

export const WRITING_WORKSPACE_TASK_DESCRIPTIONS: Record<WritingTaskType, string> = {
  task_1: 'Report / Letter (150 words, 20 minutes)',
  task_2: 'Essay (250 words, 40 minutes)',
};

// ─────────────────────────────────────────────────────────────
// Writing Evaluation types (mirrors /api/v1/writing-evaluations/*)
// ─────────────────────────────────────────────────────────────

export type EvaluationStatus = 'pending' | 'evaluated';

export type ErrorType =
  | 'Grammar'
  | 'Vocabulary'
  | 'Spelling'
  | 'Punctuation'
  | 'Sentence Structure'
  | 'Cohesion'
  | 'Repetition'
  | 'Word Choice'
  | 'Task Response';

export type ErrorSeverity = 'critical' | 'major' | 'minor';

export interface WritingError {
  id: string;
  original: string;
  error_type: ErrorType;
  explanation: string;
  correction: string;
  severity: ErrorSeverity;
  criterion: string;
  start: number;
  end: number;
  sentence?: string;
}

export interface CriterionEvaluation {
  band: number;
  label: string;
  strength: string;
  weakness: string;
  errors: string[];
  suggestions: string[];
}

export interface WritingEvaluationCriteria {
  task_response: CriterionEvaluation;
  coherence_cohesion: CriterionEvaluation;
  lexical_resource: CriterionEvaluation;
  grammatical_range_accuracy: CriterionEvaluation;
}

export interface WritingEvaluation {
  id?: string;
  submission_id: string;
  task_type: WritingTaskType;
  criteria: WritingEvaluationCriteria;
  criteria_bands: Record<string, number>;
  overall_band: number | null;
  confidence: number | null;
  is_estimate: boolean;
  word_count: number;
  source: string;
  strengths: string[];
  weaknesses: string[];
  errors: string[];
  suggestions: string[];
  evaluated_at: string | null;
  evaluation_status: EvaluationStatus;
  is_official: boolean;
  error_analysis?: WritingError[];
  mission_sync?: WritingMissionSync | null;
}

export interface WritingMissionSync {
  xp_earned: number;
  mission_completed: boolean;
  streak_bonus: number;
  predicted_band?: number | null;
  readiness_score?: number | null;
  weakest_skill?: string | null;
}

export interface WritingEvaluationSummary {
  submission_id: string;
  overall_band: number | null;
  confidence: number | null;
  word_count: number;
  task_type: WritingTaskType;
  evaluation_status: EvaluationStatus;
  created_at: string | null;
}

export interface WritingEvaluationListResponse {
  results: WritingEvaluationSummary[];
  total: number;
}

// ─────────────────────────────────────────────────────────────
// Writing Improvement Plan types ("Improve My Band")
// ─────────────────────────────────────────────────────────────

export type PlanPriority = "high" | "medium" | "low";

export type PlanSkillFocus =
  | "task_1"
  | "task_2"
  | "grammar"
  | "vocabulary"
  | "cohesion";

export interface ImprovementPlanChange {
  area: string;
  change: string;
  priority: PlanPriority;
}

export interface ImprovementPlanExercise {
  title: string;
  description: string;
  skill_focus: PlanSkillFocus;
  estimated_minutes: number;
}

export interface ImprovementPlanResource {
  title: string;
  url: string;
  why: string;
}

export interface ImprovementPlanMission {
  title: string;
  skill: string;
  sub_skill: string;
  duration_minutes: number;
  description: string;
}

export interface WritingImprovementPlan {
  id: string;
  evaluation_id: string;
  submission_id: string;
  task_type: WritingTaskType;
  current_band: number;
  target_band: number;
  band_gap: number;
  weaknesses: string[];
  current_level_description: string;
  target_level_description: string;
  specific_changes: ImprovementPlanChange[];
  practice_exercises: ImprovementPlanExercise[];
  recommended_resources: ImprovementPlanResource[];
  suggested_mission: ImprovementPlanMission;
  is_estimate: boolean;
  source: string;
  created_at: string | null;
}

export interface WritingImprovementPlanListResponse {
  results: WritingImprovementPlan[];
  total: number;
}

// ─────────────────────────────────────────────────────────────
// Writing Band Examples types ("Improve My Band — Examples")
// ─────────────────────────────────────────────────────────────

export interface ImprovedSentence {
  original: string;
  improved: string;
  explanation: string;
}

export interface VocabularyAlternative {
  from: string;
  to: string;
  why: string;
}

export interface BandExample {
  id: string;
  evaluation_id: string;
  submission_id: string;
  task_type: WritingTaskType;
  target_band: number;
  current_band: number;
  focus_areas: string[];
  key_weaknesses: string;
  improved_sentences: ImprovedSentence[];
  vocabulary_alternatives: VocabularyAlternative[];
  paragraph_structure: string;
  example_introduction: string;
  example_body_paragraph: string;
  example_conclusion: string;
  sample_answer: string | null;
  is_sample_answer: boolean;
  is_estimate: boolean;
  source: string;
  created_at: string | null;
}

export interface BandExampleListResponse {
  results: BandExample[];
  total: number;
}

// ─────────────────────────────────────────────────────────────
// Writing Reattempt Mode types
// ─────────────────────────────────────────────────────────────

export interface WritingAttempt {
  submission_id: string;
  attempt_number: number;
  overall_band: number | null;
  evaluated_at: string | null;
  bonus_xp: number;
}

export interface WritingAttemptGroup {
  attempt_group: string;
  title: string;
  task_type: WritingTaskType;
  attempts: WritingAttempt[];
  total_attempts: number;
}

export interface WritingAttemptListResponse {
  results: WritingAttemptGroup[];
  total: number;
}

export interface CriterionComparison {
  criterion: string;
  label: string;
  attempt_1_band: number;
  attempt_2_band: number;
  delta: number;
  improved: boolean;
}

export interface BandComparison {
  attempt_1: number;
  attempt_2: number;
  delta: number;
  improved: boolean;
}

export interface WordCountComparison {
  attempt_1: number;
  attempt_2: number;
  delta: number;
}

export interface TimeComparison {
  attempt_1: number;
  attempt_2: number;
  delta: number;
}

export interface WritingComparison {
  compared: boolean;
  reason?: string;
  original_submission_id: string;
  new_submission_id: string;
  overall_band: BandComparison;
  criteria: CriterionComparison[];
  word_count: WordCountComparison;
  time_seconds: TimeComparison;
  improvement: boolean;
}

export interface WritingReattemptEvaluateResponse {
  evaluation: WritingEvaluation;
  comparison: WritingComparison;
  attempt_number: number;
  bonus_xp: number;
  bonus_reason: string | null;
}

export interface WritingReattemptStartResponse {
  submission: WritingWorkspaceSubmission;
  original_submission_id: string;
  attempt_number: number;
  attempt_group: string;
  original_evaluation: WritingEvaluation;
}

// ─────────────────────────────────────────────────────────────
// Writing Coach Q&A types
// ─────────────────────────────────────────────────────────────

export interface WritingCoachMessage {
  id: string;
  conversation_id: string;
  role: "user" | "coach";
  content: string;
  structured: WritingCoachMessageStructured;
  created_at: string | null;
}

export interface WritingCoachMessageStructured {
  focus?: string;
  referenced_text?: string[];
  referenced_feedback?: string[];
  essay_text?: string;
  evaluation_id?: string;
  [key: string]: unknown;
}

export interface WritingCoachAnswer {
  conversation_id: string;
  answer: string;
  focus: string;
  referenced_text: string[];
  referenced_feedback: string[];
}

export interface WritingCoachConversation {
  id: string;
  user_id: string;
  evaluation_id: string;
  submission_id: string;
  title: string;
  status: "active" | "archived";
  messages: WritingCoachMessage[];
  created_at: string | null;
  updated_at: string | null;
}

export interface WritingCoachConversationListItem {
  id: string;
  evaluation_id: string;
  submission_id: string;
  title: string;
  status: "active" | "archived";
  message_count: number;
  last_message_at: string | null;
  created_at: string | null;
  updated_at: string | null;
}

export interface WritingCoachConversationListResponse {
  items: WritingCoachConversationListItem[];
  total: number;
  limit: number;
  offset: number;
}

// ─────────────────────────────────────────────────────────────
// Speaking Error Analysis types
// ─────────────────────────────────────────────────────────────

export type SpeakingIssueType =
  | "Grammar"
  | "Repeated Vocabulary"
  | "Weak Vocabulary"
  | "Unnatural Expression"
  | "Filler Words"
  | "Repetition"
  | "Incomplete Sentence"
  | "Hesitation Indicator"
  | "Coherence Problem"
  | "Pronunciation";

export type SpeakingSeverity = "critical" | "major" | "minor";

export type SpeakingCriterion =
  | "fluency_coherence"
  | "lexical_resource"
  | "grammatical_range"
  | "pronunciation";

export interface SpeakingErrorIssue {
  original_phrase: string;
  issue_type: SpeakingIssueType;
  explanation: string;
  why_problem: string;
  suggested_improvement: string;
  criterion_affected: SpeakingCriterion;
  severity: SpeakingSeverity;
  context?: string;
}

export interface SpeakingErrorAnalysis {
  id: string;
  response_id: string;
  part: string;
  topic: string;
  issues: SpeakingErrorIssue[];
  overall_band: number;
  fluency_coherence_band: number;
  lexical_resource_band: number;
  grammatical_range_band: number;
  pronunciation_band: number;
  feedback: string;
  issue_count: number;
  high_severity_count: number;
  medium_severity_count: number;
  low_severity_count: number;
  is_estimate: boolean;
  source: string;
  created_at: string | null;
}

export interface SpeakingErrorAnalysisListResponse {
  results: SpeakingErrorAnalysis[];
  total: number;
}

File: backend/app/models/writing_workspace.py
"""
Pydantic schemas for the Writing Workspace API.
"""
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


# ─── Submission CRUD ──────────────────────────────────────────────────
class SubmissionStart(BaseModel):
    """Request to start/resume a writing submission."""
    prompt_id: str


class SubmissionSave(BaseModel):
    """Request to auto-save a draft."""
    essay_text: str
    time_seconds_spent: Optional[int] = None


class SubmissionSubmit(BaseModel):
    """Request to submit a draft for evaluation."""
    time_seconds_spent: Optional[int] = None


class PromptResponse(BaseModel):
    """A single writing prompt."""
    id: str
    task_type: str
    title: str
    prompt_text: str
    word_limit: int
    time_limit_seconds: int
    difficulty: int
    topics: List[str] = []


class PromptsResponse(BaseModel):
    """List of writing prompts."""
    task_type: str
    prompts: List[PromptResponse]
    total: int


class WritingSubmissionSummary(BaseModel):
    """Pre-submission summary captured at submit time."""
    word_count: int
    word_limit: int
    time_seconds_spent: int
    time_limit_seconds: int
    meets_word_requirement: bool
    within_time_limit: bool
    warnings: List[str] = []
    submitted_at: Optional[str] = None


class SubmissionResponse(BaseModel):
    """A writing workspace submission."""
    id: str
    user_id: str
    prompt_id: Optional[str] = None
    task_type: str
    title: str
    prompt_text: Optional[str] = None
    word_limit: int
    time_limit_seconds: int
    essay_text: str
    word_count: int
    time_seconds_spent: int
    status: str
    is_locked: bool
    evaluation_status: Optional[str] = None
    submission_summary: Dict[str, Any] = {}
    created_at: Optional[str] = None
    updated_at: Optional[str] = None
    submitted_at: Optional[str] = None


class SubmissionListResponse(BaseModel):
    """List of submissions."""
    results: List[SubmissionResponse]
    total: int


# ─── Evaluation ───────────────────────────────────────────────────────
class WritingError(BaseModel):
    """
    A single detected issue within the essay.

    Carries the original text, category, explanation, suggested correction,
    severity, the IELTS criterion affected, and character offsets used by the
    UI to highlight the problematic section in the essay.

    ``correction`` fixes only this one issue — the essay is never rewritten
    automatically as a whole.
    """
    id: str
    original: str
    error_type: str
    explanation: str
    correction: str
    severity: str = "minor"  # critical | major | minor
    criterion: str = "grammatical_range_accuracy"
    start: int = 0
    end: int = 0
    sentence: str = ""


class CriterionEvaluation(BaseModel):
    """A single criterion evaluation."""
    band: float
    label: str
    strength: str
    weakness: str
    errors: List[str] = []
    suggestions: List[str] = []


class WritingCriteriaEvaluation(BaseModel):
    """All four criteria evaluations."""
    task_response: CriterionEvaluation
    coherence_cohesion: CriterionEvaluation
    lexical_resource: CriterionEvaluation
    grammatical_range_accuracy: CriterionEvaluation


class WritingEvaluationResponse(BaseModel):
    """
    Full evaluation response returned to the client.

    AI scoring is not implemented in this phase — pending records carry no
    bands and ``evaluation_status`` is 'pending'.  A future phase fills the
    criteria/bands and flips the status to 'evaluated'.
    """
    task_type: str
    criteria_bands: Dict[str, Any] = {}
    criteria_detail: Dict[str, Any] = {}
    overall_band: Optional[float] = None
    confidence: Optional[float] = None
    is_estimate: bool = True
    word_count: int = 0
    source: str = "pending"
    strengths: Optional[List[str]] = None
    weaknesses: Optional[List[str]] = None
    errors: Optional[List[str]] = None
    suggestions: Optional[List[str]] = None
    evaluated_at: Optional[str] = None
    evaluation_status: str = "pending"
    is_official: bool = False
    error_analysis: Optional[List[WritingError]] = None
    mission_sync: Optional[Dict[str, Any]] = None
    attempt_number: int = 1


class WritingEvaluationSummary(BaseModel):
    """Lightweight evaluation summary for lists."""
    submission_id: str
    overall_band: float
    confidence: float
    word_count: int
    task_type: str
    created_at: Optional[str] = None


class WritingEvaluationListResponse(BaseModel):
    """List of evaluation summaries."""
    results: List[WritingEvaluationSummary]
    total: int


class ImprovementPlanChange(BaseModel):
    """A single concrete change recommended to the student."""
    area: str
    change: str
    priority: str  # high | medium | low


class ImprovementPlanExercise(BaseModel):
    """A single practice exercise in the plan."""
    title: str
    description: str
    skill_focus: str
    estimated_minutes: int


class ImprovementPlanResource(BaseModel):
    """A recommended resource in the plan."""
    title: str
    url: str
    why: str


class ImprovementPlanMission(BaseModel):
    """A suggested mission to add to the study plan."""
    title: str
    skill: str
    sub_skill: str
    duration_minutes: int
    description: str


class WritingImprovementPlanResponse(BaseModel):
    """Full improvement plan response returned to the client."""
    id: str
    evaluation_id: str
    submission_id: str
    task_type: str
    current_band: float
    target_band: float
    band_gap: float
    weaknesses: List[str] = []
    current_level_description: str
    target_level_description: str
    specific_changes: List[ImprovementPlanChange] = []
    practice_exercises: List[ImprovementPlanExercise] = []
    recommended_resources: List[ImprovementPlanResource] = []
    suggested_mission: ImprovementPlanMission = None  # type: ignore[assignment]
    is_estimate: bool = True
    source: str = "ai"
    created_at: Optional[str] = None


class WritingImprovementPlanListResponse(BaseModel):
    """List of improvement plan summaries."""
    results: List[WritingImprovementPlanResponse]
    total: int


class ImprovedSentence(BaseModel):
    """A sentence-level improvement example."""
    original: str
    improved: str
    explanation: str


class VocabularyAlternative(BaseModel):
    """A vocabulary alternative suggestion."""
    from_: str = Field(alias="from")
    to: str
    why: str

    model_config = {"populate_by_name": True}


class BandExampleResponse(BaseModel):
    """Full band-level example response returned to the client."""
    id: str
    evaluation_id: str
    submission_id: str
    task_type: str
    target_band: float
    current_band: float
    focus_areas: List[str] = []

    key_weaknesses: str
    improved_sentences: List[ImprovedSentence] = []
    vocabulary_alternatives: List[VocabularyAlternative] = []
    paragraph_structure: str
    example_introduction: str
    example_body_paragraph: str
    example_conclusion: str
    sample_answer: Optional[str] = None
    is_sample_answer: bool = False

    is_estimate: bool = True
    source: str = "ai"
    created_at: Optional[str] = None


class BandExampleListResponse(BaseModel):
    """List of band examples."""
    results: List[BandExampleResponse]
    total: int


class SpeakingErrorIssue(BaseModel):
    """A single speaking error/issue found in a transcript."""
    original_phrase: str
    issue_type: str
    explanation: str
    why_problem: str
    suggested_improvement: str
    criterion_affected: str
    severity: str
    context: Optional[str] = None


class SpeakingErrorAnalysisResponse(BaseModel):
    """Full speaking error analysis response."""
    id: str
    response_id: str
    part: str
    topic: str
    issues: List[SpeakingErrorIssue] = Field(default_factory=list)
    overall_band: float
    fluency_coherence_band: float
    lexical_resource_band: float
    grammatical_range_band: float
    pronunciation_band: float
    feedback: str
    issue_count: int
    high_severity_count: int
    medium_severity_count: int
    low_severity_count: int
    is_estimate: bool = True
    source: str = "ai"
    created_at: Optional[str] = None


class SpeakingErrorAnalysisListResponse(BaseModel):
    """List of speaking error analyses."""
    results: List[SpeakingErrorAnalysisResponse]
    total: int


class SpeakingSpecificChange(BaseModel):
    area: str
    change: str
    priority: str

class SpeakingPracticeExercise(BaseModel):
    title: str
    description: str
    skill_focus: str
    estimated_minutes: int

class SpeakingResource(BaseModel):
    title: str
    url: str
    why: str

class SpeakingMission(BaseModel):
    title: str
    skill: str = "speaking"
    sub_skill: str = ""
    duration_minutes: int = 15
    description: str = ""

class SpeakingImprovementPlanResponse(BaseModel):
    id: str
    response_id: str
    current_band: float
    target_band: float
    band_gap: float
    strongest_criterion: str
    weakest_criterion: str
    criterion_priorities: dict = Field(default_factory=dict)
    current_level_description: str
    target_level_description: str
    specific_changes: List[SpeakingSpecificChange] = Field(default_factory=list)
    practice_exercises: List[SpeakingPracticeExercise] = Field(default_factory=list)
    practice_topics: List[str] = Field(default_factory=list)
    recommended_resources: List[SpeakingResource] = Field(default_factory=list)
    suggested_daily_minutes: int = 15
    next_speaking_task: str
    suggested_mission: SpeakingMission = Field(default_factory=SpeakingMission)
    is_estimate: bool = True
    source: str = "ai"
    created_at: Optional[str] = None

class SpeakingImprovementPlanListResponse(BaseModel):
    results: List[SpeakingImprovementPlanResponse]
    total: int
